contracts/vesting/src/vesting.rs

```diff
@@ -0,0 +1,559 @@
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+        querier: Option<QuerierWrapper>,
+    ) -> Result<(), ContractError> {
+        if let Some(querier) = querier {
+            let msg = MixnetExecuteMsg::UnDelegateFromMixnodeOnBehalf {
+                mix_identity: mix_identity.clone(),
+                delegate_addr: self.address.clone(),
+            };
+            querier.query_wasm_smart(DEFAULT_MIXNET_CONTRACT_ADDRESS, &msg)?;
+        }
+        self.track_undelegation(mix_identity, storage)?;
+        Ok(())
+    }
+
+    fn track_delegation(
+        &self,
+        block_time: Timestamp,
+        mix_identity: IdentityKey,
+        delegation_amount: Coin,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        let mut delegations =
+            if let Some(delegations) = get_account_delegations(storage, &self.address) {
+                delegations
+            } else {
+                Vec::new()
+            };
+        delegations.push(DelegationData {
+            mix_identity,
+            amount: delegation_amount.amount,
+            block_time,
+        });
+        set_account_delegations(storage, &self.address, delegations)?;
+        Ok(())
+    }
+
+    fn track_undelegation(
+        &self,
+        mix_identity: IdentityKey,
+        storage: &mut dyn Storage,
+    ) -> Result<(), ContractError> {
+        // This has to exist in storage at this point.
+        let delegations = get_account_delegations(storage, &self.address)
+            .unwrap()
+            .into_iter()
+            .filter(|d| d.mix_identity != mix_identity)
+            .collect();
+        // Since we're always removing the entire delegation we can just drop the key
+
+        Ok(set_account_delegations(
+            storage,
+            &self.address,
+            delegations,
+        )?)
+    }
+}
+
+impl VestingAccount for PeriodicVestingAccount {
+    fn locked_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        // Returns 0 in case of underflow.
+        Coin {
+            amount: Uint128(
+                self.get_vesting_coins(block_time, env)
+                    .amount
+                    .u128()
+                    .saturating_sub(
+                        self.get_delegated_vesting(block_time, env, storage)
+                            .amount
+                            .u128(),
+                    ),
+            ),
+            denom: DENOM.to_string(),
+        }
+    }
+
+    fn spendable_coins(
+        &self,
+        block_time: Option<Timestamp>,
+        env: &Env,
+        storage: &dyn Storage,
+    ) -> Coin {
+        Coin {
+            amount: Uint128(
+                self.get_balance(storage)
+                    .u128()
+                    .saturating_sub(self.locked_coins(block_time, env, storage).amount.u128()),
```

Comment:
Again, why is it a `saturating_sub`. Could it ever fail?

Reply:
Also, in some other location you said that spendable balance could be negative. However, this code seems to contradict it as you can't store negative values in the `Coin`

Reply:
`saturating_sub` will never fail, in case of an underflow it goes to `0`. The balance itself can be negative, but I don't think we care if its negative or 0, in either case you have nothing to spend.